django/tasks/base.py

```diff
@@ -0,0 +1,310 @@
+            )
+
+        return result
+
+    async def aget_result(self, result_id):
+        """
+        Retrieve the result for a task of this type by its id (if one exists).
+        If one doesn't, or is the wrong type, raises ResultDoesNotExist.
+        """
+        result = await self.get_backend().aget_result(result_id)
+
+        if result.task.func != self.func:
+            raise TaskIntegrityError(
+                f"Task does not match (received {result.task.module_path!r})"
+            )
+
+        return result
+
+    def call(self, *args, **kwargs):
+        if iscoroutinefunction(self.func):
+            return async_to_sync(self.func)(*args, **kwargs)
+        return self.func(*args, **kwargs)
+
+    async def acall(self, *args, **kwargs):
+        if iscoroutinefunction(self.func):
+            return await self.func(*args, **kwargs)
+        return await sync_to_async(self.func)(*args, **kwargs)
+
+    def get_backend(self):
+        from . import tasks
+
+        return tasks[self.backend]
+
+    @property
+    def module_path(self):
+        return f"{self.func.__module__}.{self.func.__qualname__}"
+
+
+def task(
+    function=None,
+    *,
+    priority=DEFAULT_PRIORITY,
+    queue_name=DEFAULT_QUEUE_NAME,
+    backend=DEFAULT_TASK_BACKEND_ALIAS,
+    enqueue_on_commit=None,
+    takes_context=False,
+):
+    """
+    A decorator used to create a task.
+    """
+    from . import tasks
+
+    def wrapper(f):
+        return tasks[backend].task_class(
+            priority=priority,
+            func=f,
+            queue_name=queue_name,
+            backend=backend,
+            enqueue_on_commit=enqueue_on_commit,
+            takes_context=takes_context,
+        )
+
+    if function:
+        return wrapper(function)
+
+    return wrapper
+
+
+@dataclass(frozen=True, slots=True, kw_only=True)
+class TaskError:
+    exception_class_path: str
+    traceback: str
+
+    @property
+    def exception_class(self):
+        # Lazy resolve the exception class
+        exception_class = import_string(self.exception_class_path)
+
+        if not isclass(exception_class) or not issubclass(
+            exception_class, BaseException
+        ):
+            raise ValueError(
+                f"{self.exception_class_path!r} does not reference a valid exception."
+            )
+
+        return exception_class
+
+
+@dataclass(frozen=True, slots=True, kw_only=True)
+class TaskResult:
+    task: Task
+    """The task for which this is a result"""
+
+    id: str
+    """A unique identifier for the task result"""
+
+    status: ResultStatus
+    """The status of the running task"""
+
+    enqueued_at: Optional[datetime]
+    """The time this task was enqueued"""
+
+    started_at: Optional[datetime]
+    """The time this task was started"""
+
+    finished_at: Optional[datetime]
+    """The time this task was finished"""
+
+    last_attempted_at: Optional[datetime]
+    """The time this task was last attempted to be run"""
```

Comment:
I am really tempted for us to remove this because I _believe_ this is misleading.
This will always either be `None` or the same as `started_at` but it _feels_ like it might give you when the previous attempt started.
Either I would want us to change this or remove it entirely. If it does what I said it feels like it should do, we need better testing showing this

Reply:
With the current implementation, yes it's exactly the same as `started_at`, since multiple attempts isn't currently supported. I intentionally designed the shape of the data now to support multiple attempts, so it's slightly less breaking when new fields are added.

Happy to drop it through if you think it's not worth it?

Reply:
Does the current system (without multiple attempts) support an attempt that does not actually start?

I'd keep it there from the beginning of the plan is to support multiple attempts.